pkg/kn/commands/plugin/path_verifier.go

```diff
@@ -0,0 +1,102 @@
+// Copyright © 2018 The Knative Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plugin
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+// PathVerifier receives a path and determines if it is valid or not
+type PathVerifier interface {
+	// Verify determines if a given path is valid
+	Verify(path string) []error
+}
+
+// CommandOverrideVerifier verifies that existing kn commands are not overriden
+type CommandOverrideVerifier struct {
+	Root        *cobra.Command
+	SeenPlugins map[string]string
+}
+
+// Verify implements PathVerifier and determines if a given path
+// is valid depending on whether or not it overwrites an existing
+// kn command path, or a previously seen plugin.
+func (v *CommandOverrideVerifier) Verify(path string) []error {
+	if v.Root == nil {
+		return []error{fmt.Errorf("unable to verify path with nil root")}
+	}
+
+	// extract the plugin binary name
+	segs := strings.Split(path, "/")
+	binName := segs[len(segs)-1]
+
+	cmdPath := strings.Split(binName, "-")
+	if len(cmdPath) > 1 {
+		// the first argument is always "kn" for a plugin binary
+		cmdPath = cmdPath[1:]
+	}
+
+	errors := []error{}
+	isExec, err := isExecutable(path)
+	if err == nil && !isExec {
+		errors = append(errors, fmt.Errorf("warning: %s identified as a kn plugin, but it is not executable", path))
+	} else if err != nil {
+		errors = append(errors, fmt.Errorf("error: unable to identify %s as an executable file: %v", path, err))
+	}
+
+	if existingPath, ok := v.SeenPlugins[binName]; ok {
+		errors = append(errors, fmt.Errorf("warning: %s is overshadowed by a similarly named plugin: %s", path, existingPath))
+	} else {
+		v.SeenPlugins[binName] = path
+	}
+
+	cmd, _, err := v.Root.Find(cmdPath)
```

Comment:
I've mentioned it elsewhere already why this check here is not sufficient. Let me try to explain it with an example:

* Assume we already have a command `kn importer list-types`, i.e. a command "importer" and a sub-type "list-types", built-in into kn.
* Now kn finds a plugin called `kn-importer-list_types`. In this case the array `cmdPath` will contain `("importer", "list_types")` (note, that `_` is not converted to a `-`). So `v.Root.Find` will fail and hence verify succeed.

Or did I miss something ?

Reply:
```bash
➜  client git:(kn-plugins4) cat /usr/local/bin/kn-foo-bar
#!/bin/bash

echo "My first command-line argument was '$1'"
➜  client git:(kn-plugins4) cat /usr/local/bin/kn-foo_bar2
#!/bin/bash

echo "My plugin got command-line argument was '$1'"
➜  client git:(kn-plugins4) ./kn plugin list
The following compatible plugins are available:

/usr/local/bin/kn-foo-bar
/usr/local/bin/kn-foo_bar2
➜  client git:(kn-plugins4) ./kn foo bar lol
My first command-line argument was 'lol'
➜  client git:(kn-plugins4) ./kn foo_bar2 lol
My plugin got command-line argument was 'lol'
```

Reply:
Apologies, I mis-interpreted that line (confused 'old' with 'new'):

https://github.com/maximilien/client/blob/1819e52821d36202a2e070bdcf531f1fbfd74c93/pkg/kn/commands/plugin/plugin_handler.go#L80

What is it supposed to do then when not escaping _ with a - for allowing dashed command? (I really thought that the replacement was vice versa which in my naive thinking makes more sense, therefore that brain slip probably)

Or let's pose the question around: How do you allow dashed commands?

Reply:
Yup, that's a known problem. Not allowed now. 

But if we must have I guess we'd need some convention, naively: `/usr/local/bin/kn-foo--bar` should be interpreted as plugin named: `kn foo-bar`

Thoughts?

Let me open a separate issue on this since we can discuss there and implement once resolved.

Reply:
@maximilien if you look to [kubectl documentation](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/#names-with-dashes-and-underscores) they allow `-` in the command as well and treat it on the same footprint as `_`:

> Although the kubectl plugin mechanism uses the dash (-) in plugin filenames to separate the sequence of sub-commands processed by the plugin, it is still possible to create a plugin command containing dashes in its commandline invocation by using underscores (_) in its filename.
> ....
> Note that the introduction of underscores to a plugin filename does not prevent us from having commands such as kubectl foo_bar. The command from the above example, can be invoked using either a dash (-) or an underscore (_)

I think we can do the same and shouldn't be that hard. @maximilien in order to copy the similarity to kubectl that you already emphasized, we should provide the same semantics and allow dashes (i.e. treat a `-` and `_` in the file name both as `-` and `_`). We should not introduce a new convention. 'hope that aligns with your thinking to stay close to kubectl in that respect (plugin feature).

BTW, this diff in behaviour was really what confused me. Apologies for that.

Reply:
Resolving since issue #272 captures and should be used for future discussion on this point.

Reply:
All for not introducing new convention and tailing on `kubectl` plugins. Hence why initial PR is specifically trying to be like `kubectl` plugins. Heck, in the name of this PR :)

Reply:
Ok, if we then agree that we are currently not like kubectl PRs. 

Btw, I think it would be honest if the spots where code has been copied over would be marked with a comment or so. Not only for credits, but also to have a backlink in cases things are changes upstream.